### PR TITLE
interactive selection of groups

### PR DESCRIPTION
**Pull request by danielpastor97**
## Description
This PR implements an interactive selection of the groups that are going to be used for the calculation of the contacts (issue #9). The idea is quite similar to the make_ndx command of gromacs and the greatest advantage is that can be used for getting the contacts at any level (molecule, residue, atom) and between any two groups (not only lipid-protein).

## Features
Notable points that this PR has either accomplished:
  - [x] You can get into the interactive mode for the selection of the groups simply by typing `-i` or `-i True` when you are calling `ufcc` from the command-line.
  - [x] A list of groups is created by default where the groups `Database` and `Query` are the first two, and are the ones that will be used for the calculation of the contacts, so there are key actions that will allow the user to update these groups. Other than `Database` and `Query`, the list of groups included by default are: `System`, `Protein`, `Lipids`, and a group for any other resname included in `System` and not included in `Protein`, that can be for example: POPC, DOPC, POPE, SOL, SOD, CLA, LIG.
  - [x] You can update the `Database` and `Query` groups using the key actions `db` and `qr`, followed by the identifier of the group that you want to select as database or query for the calculation of the contacts.
  - [x] You can create new groups starting from the initial groups using two key actions: i) `gb` creates subgroups grouped by any topological attribute as for example the resnames, so you can do `gb 2 resnames`, where 2 is the identifier for the `Protein` group, and it will create a new group for each resname in the `Protein` group; and ii) `sl` creates new groups by splitting a previous group at the desired level ("segment", "residue", "atom", "molecule").
  - [x] You can merge two or more groups using the `add` key action as `add Group_ID1 Group_ID2 ... Group_IDn`, which creates a new group for the combination of the selected groups.
  - [x] Delete groups using `del Group_ID`.
  - [x] Print a list of the current groups with `lg`.
  - [x] Print a list of key actions using `h` (from help).
  - [x] Exit the interactive mode with `e`.
  
## Todos
Features to be added or that need to be resolved in order to merge the pull request.
- [x] Raise an error message if the structure of the command for each key actions does not match the expected structure.

## Status
- [x] Ready to go

### CODE REVIEW

File: bin/ufcc [hunk 1]
```diff
@@ -63,6 +63,91 @@ with warnings.catch_warnings(record=True) as w:
...
+qr  :  update query group for contacts calculation. (i.e. >> d Group_ID) where Group_ID is the identifier of the group on the list of groups.
+gp :  create subgroups grouped by any topology attribute ("names", "resnames", "masses", etc). (i.e. >> gp Group_ID resnames) creates new groups for every different resname in the Group_ID selected.
+sl  :  create new groups by splitting a previous group at the desired level ("segment", "residue", "atom", "molecule"). (i.e. >> sl Group_ID residue) splits the Group_ID selected into all the residues that make it up, and create a new group for each of them.
+add:  merge two or more groups and create a new group with the combination. (i.e. >> add Group_ID1 Group_ID2 ... Group_IDn) creates a new group for the combination of the groups Group_ID1 Group_ID2 ... Group_IDn.
+del:  delete a group from the list of groups. (i.e. >> del Group_ID) remove the Group_ID selected.
+lg :  print the list of groups.
+h  :  print the help for all the available action keys.
+e  :  exit interactive selection and calculate de contacts between the Query and Database groups.
```
Comment (bisejdiu) [inline bin/ufcc:90]: This is excellent! 

I can see a few ways we can improve this, however. One idea is to add support for `+` and `-`, which can be mapped into set operations (e.g. `+` can be `union`, `|` can be `intersection`, `-` can be `A - (A∩B)`, and others). 

This is not difficult to do, have a look at this: https://www.codingem.com/python-__add__-method/

But also note that from my understanding, MDAnalysis already implements many of these. I checked the code and it should be possible to add atomgroups directly (and do other operations as well): [link](https://github.com/MDAnalysis/mdanalysis/blob/b302e40ee45831409b5641311ac51fc4db08e07a/package/MDAnalysis/core/groups.py#L2394-L2448)

File: bin/ufcc [hunk 2]
```diff
@@ -63,6 +63,91 @@ with warnings.catch_warnings(record=True) as w:
         print(mass_warn)
         warn2report.append(mass_warn)
 
+# For interactive selection of the groups for the contacts calculation
+if args.i_bool:
+    groups_dict = {
```
Comment (bisejdiu) [inline bin/ufcc:68]: You should consider putting the code into a separate module so it can have its own API and be called outside of the command line.

**Review by bisejdiu — CHANGES_REQUESTED** [review]

File: bin/ufcc [hunk 3]
```diff
@@ -7,6 +7,7 @@ r"""Argument parser to use ufcc from the command-line
 """
 
 import argparse
+from tkinter import N
```
[the remark this answers is not in the document]
Comment (danielpastor97) [reply bin/ufcc:10]: Completely, we don't need tkinter for anything, It must be a mistake with the auto-completion, I will remove it right away.

**Comment by bisejdiu** [comment]
@danielpastor97 Pressing Return/Enter without providing input gives an IndexError.

File: bin/ufcc [hunk 4]
```diff
@@ -63,6 +63,91 @@ with warnings.catch_warnings(record=True) as w:
         print(mass_warn)
         warn2report.append(mass_warn)
 
+# For interactive selection of the groups for the contacts calculation
+if args.i_bool:
+    groups_dict = {
+        0: ['Database', target_system.database.selected],
+        1: ['Query', target_system.query.selected],
```
[the remark this answers is not in the document]
Comment (danielpastor97) [reply bin/ufcc]: I put them apart and will appear to the user as:
Screen Shot 2022-07-25 at 11 07 19 AM

File: bin/ufcc (hunk 2 above)
Comment (danielpastor97) [reply bin/ufcc]: Done! I isolated the code for the interactive selection in a function in the `ufcc.interactive_sel` module.
You can use it simply by calling this function as in:
```python
from ufcc import UFCC
from ufcc.interactive_sel import interactive_selection
from ufcc.sampledata import GIRK

target_system = UFCC(GIRK.coordinates, GIRK.trajectory)
target_system = interactive_selection(target_system)
```

**Comment by danielpastor97** [comment]
Hey Besian, have you had the chance to try this out? Please let me know if you think that there is something left that needs to be done in order to close it.

**Review by bisejdiu — APPROVED** [review]

**Pull request merged**